File: src/viewer/controllers/model/CharacterClassOptionsController.ts
import { Controller } from "../../../core/controllers/Controller";
import { ModelStore } from "../../stores/ModelStore";
import { range } from "lodash";
import { Property } from "../../../core/observable/property/Property";
import { SectionId } from "../../../core/model";

export class CharacterClassOptionsController extends Controller {
    readonly enabled: Property<boolean>;
    readonly current_section_id: Property<SectionId>;
    readonly current_body_options: Property<readonly number[]>;
    readonly current_body: Property<number | undefined>;

    constructor(private readonly store: ModelStore) {
        super();

        this.enabled = store.current_character_class.map(cc => cc != undefined);

        this.current_section_id = store.current_section_id;

        this.current_body_options = store.current_character_class.map(character_class =>
            character_class ? range(1, character_class.body_style_count + 1) : [],
        );

        this.current_body = store.current_body.map(body =>
            body == undefined ? undefined : body + 1,
        );
    }

    set_current_section_id = (section_id?: SectionId): void => {
        if (section_id != undefined) {
            this.store.set_current_section_id(section_id);
        }
    };

    set_current_body = (body?: number): void => {
        this.store.set_current_body(body == undefined ? undefined : body - 1);
    };
}

File: web/src/main/kotlin/world/phantasmal/web/viewer/controllers/CharacterClassOptionsController.kt
package world.phantasmal.web.viewer.controllers

import world.phantasmal.observable.cell.Cell
import world.phantasmal.observable.cell.isNotNull
import world.phantasmal.observable.cell.map
import world.phantasmal.observable.cell.plus
import world.phantasmal.web.shared.dto.SectionId
import world.phantasmal.web.viewer.stores.ViewerStore
import world.phantasmal.webui.controllers.Controller

class CharacterClassOptionsController(private val store: ViewerStore) : Controller() {
    val enabled = store.currentCharacterClass.isNotNull()
    val currentSectionId: Cell<SectionId> = store.currentSectionId
    val currentBodyOptions: Cell<List<Int>> = store.currentCharacterClass.map { char ->
        if (char == null) emptyList() else (1..char.bodyStyleCount).toList()
    }
    val currentBody: Cell<Int> = store.currentBody + 1

    suspend fun setCurrentSectionId(sectionId: SectionId) {
        store.setCurrentSectionId(sectionId)
    }

    suspend fun setCurrentBody(body: Int) {
        store.setCurrentBody(body - 1)
    }
}
